### ISSUE TITLE
[C#] Implement new Concept Exercise: attribute

### ISSUE THREAD

**Issue by ErikSchierboom** [exercism/v3#1176]
This issue describes how to implement the `attributes` concept exercise for the C# track.

## Getting started

**Please please please read the docs before starting.** Posting PRs without reading these docs will be a lot more frustrating for you during the review cycle, and exhaust Exercism's maintainers' time. So, before diving into the implementation, please read up on the following documents:

- [The features of v3](https://github.com/exercism/v3/blob/master/docs/concept-exercises.md).
- [Rationale for v3](https://github.com/exercism/v3/blob/master/docs/rationale-for-v3.md).
- [What are concept exercise and how they are structured?](https://github.com/exercism/v3/blob/master/docs/features-of-v3.md)

## Goal

The goal of this exercise is to teach the student the basics of the Concept of Attributes in C#.

## Learning objectives

- Know what attributes are.
- Know the limitations of attributes.
- Know how to add attributes to values.
- Creating a custom attribute.
- Using the `AttributeUsageAttribute` to restrict usage.

## Out of scope

- Using reflection to read attributes.

## Concepts

The Concepts this exercise unlocks are:

- `attributes`: know what attributes are; know the limitations of attributes; know how to add attributes to values; creating a custom attribute.

## Prequisites

This exercise's prerequisites Concepts are:

- `inheritance`: know how to inherit from a class.

Any data types used in this exercise (e.g. `strings`) should also be added as prerequisites.

### Hints

- [Attributes][attributes]: what attributes are and how to use them.
- [Creating custom attributes][custom-attributes]: how to create custom attributes.
- [Restricting attribute usage][attributeusage]: how to restrict attribute usage.

### After

- [Attributes][attributes]: what attributes are and how to use them.
- [Creating custom attributes][custom-attributes]: how to create custom attributes.
- [Restricting attribute usage][attributeusage]: how to restrict attribute usage.

## Representer

This exercise does not require any specific representation logic to be added to the [representer][representer].

## Analyzer

This exercise does not require any specific analyzer logic to be added to the [analyzer][analyzer].

## Implementing

If you'd like to work on implementing this exercise, the first step is to let us know through a comment on this issue, to prevent multiple people from working on the same exercise. If you have any questions while implementing the exercise, please also post them as comments in this issue.

Implementing the exercise means creating the following files:

<pre>
languages
└── csharp
    └── exercises
        └── concept
            └── attributes
                ├── .docs
                |   ├── after.md
                |   ├── hints.md
                |   ├── instructions.md
                |   └── introduction.md
                ├── .meta
                |   ├── config.json
                |   ├── design.md
                |   └── Example.cs
                ├── Attributes.csproj
                ├── Attributes.cs
                └── AttributesTests.cs
</pre>

## Step 1: add .docs/introduction.md

This file contains an introduction to the concept. It should make the exercise's learning goals explicit and provide a short introduction with enough detail for the student to complete the exercise. The aim is to give the student just enough context to figure out the solution themselves, as research has shown that self-discovery is the most effective learning experience. Using the proper technical terms in the descriptions will be helpful if the student wants to search for more information. If the exercise introduces new syntax, an example of the syntax should always be included; students should not need to search the web for examples of syntax.

As an example, the introduction to a "strings" exercise might describe a string as just a "Sequence of Unicode characters" or a "series of bytes". Unless the student needs to understand more nuanced details in order to solve the exercise, this type of brief explanation (along with an example of its syntax) should be sufficient information for the student to solve the exercise.

## Step 2: add .docs/instructions.md

This file contains instructions for the exercise. It should explicitly explain what the student needs to do (define a method with the signature `X(...)` that takes an A and returns a Z), and provide at least one example usage of that function. If there are multiple tasks within the exercise, it should provide an example of each.

## Step 3: add .docs/hints.md

If the student gets stuck, we will allow them to click a button requesting a hint, which shows this file. This will not be a "recommended" path and we will (softly) discourage them using it unless they can't progress without it. As such, it's worth considering that the student reading it will be a little confused/overwhelmed and maybe frustrated.

The file should contain both general and task-specific "hints". These hints should be enough to unblock almost any student. They might link to the docs of the functions that need to be used.

The hints should not spell out the solution, but instead point to a resource describing the solution (e.g. linking to documentation for the function to use).

## Step 4: add .docs/after.md

Once the student completes the exercise they will be shown this file, which should provide them with a summary of what the exercise aimed to teach. This document can also link to any additional resources that might be interesting to the student in the context of the exercise.

The above four files are also all described in the [concept exercises document][concept-exercises].

## Step 5: update languages/csharp/config.json

An entry should be added to the track's `config.json` file for the new concept exercise:

```json
{
  ...
  "exercises": {
    "concept": [
      ...
      {
        "slug": "attributes",
        "uuid": "29ea82eb-3874-4a0b-974c-440093497cc7",
        "concepts": ["attributes"],
        "prerequisites": ["inheritance"]
      }
    ]
  }
}
```

## Step 6: adding track-specific files

These files are specific to the C# track:

- `Attributes.csproj`: the C# project file. Note: the C# 8 nullabillity checking should be enabled.
- `AttributesTests.cs`: the test suite.
- `Attributes.cs`. the stub implementation file, which is the starting point for students to work on the exercise.
- `.meta/Example.cs`: an example implementation that passes all the tests.

Check out [an existing exercise][exercise-example] to see what these files should look like.

## Step 7: update the general concept document

Add the exercise to the [concept's shared document's][reference-doc] `## Implementations` section ([example][reference-example]).

## Step 8: updating char of implemented exercises

- Add the exercise to the [list of implemented exercises][implemented-exercises].

## Step 9: add .meta/design.md:

This file contains information on the exercise's design, which includes things like its goal, its teaching goals, what not to teach, and more ([example][design-example]). This information can be extracted from this GitHub issue.

## Step 10: add .meta/config.json:

This file contains meta information on the exercise, which currently only includes the exercise's contributors ([example][config.json-example]).

### Inspiration

When implementing this exericse, it can be very useful to look at [already implemented C# exercises][implemented-exercises]. You can also check the [general char concept documentation][reference-doc] to see if any other languages have already implemented a attributes exercise.

[how-to-implement-a-concept-exercise]: https://github.com/exercism/v3/blob/master/docs/maintainers/generic-how-to-implement-a-concept-exercise.md
[implemented-exercises]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/README.md#implemented-exercises
[reference]: https://github.com/exercism/v3/blob/master/languages/csharp/reference/README.md#reference-docs
[reference-doc]: https://github.com/exercism/v3/blob/master/reference/concepts/pattern_matching.md
[reference-example]: https://github.com/exercism/v3/blob/master/reference/types/string.md#implementations
[analyzer]: https://github.com/exercism/csharp-analyzer
[representer]: https://github.com/exercism/csharp-representer
[exercise-example]: https://github.com/exercism/v3/tree/master/languages/csharp/exercises/concept/numbers-floating-point
[design-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/design.md
[config.json-example]: https://github.com/exercism/v3/blob/master/languages/csharp/exercises/concept/numbers/.meta/config.json
[concept-exercises]: https://github.com/exercism/v3/blob/master/docs/concept-exercises.md
[attributes]: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/attributes/
[custom-attributes]: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/attributes/creating-custom-attributes
[attributeusage]: https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/concepts/attributes/attributeusage

**Comment by mikedamay** [comment]
> Using reflection to read attributes.

I can't help feeling that if we don't allow attributes to be read not only will custom attributes seem rather pointless but the compiler and framework provided ones as well.  We might have to pretty much give the code to access them but at least students would see the point of them.

**Comment by ErikSchierboom** [comment]
> I can't help feeling that if we don't allow attributes to be read not only will custom attributes seem rather pointless but the compiler and framework provided ones as well. We might have to pretty much give the code to access them but at least students would see the point of them.

Maybe we should have `reflection` as a prerequisite of this exercise? That way, we _can_ use reflection here. I'd be hesitant to also introduce reflection here, or do you think that could work well?

**Comment by mikedamay** [comment]
> Maybe we should have reflection as a prerequisite of this exercise? 

The trade-off is complexity (of bringing in reflection either as part of exercise or a prerequisite) vs. leaving a bit of a mystery about what the point of attributes is.  I too favour for having a separate reflection exercise.  Our audience is seasoned programmers who should be able to handle plus it will give us an opportunity to tell those dynamic coder types that it exists to use it sparingly.

**Comment by ErikSchierboom** [comment]
Excellent!
